src/Effects/RAVU/ravu-r2.hlsl

```diff
@@ -0,0 +1,510 @@
+// Generated by ravu.py
+// 
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+// 
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+
+
+//!MAGPIE EFFECT
+//!VERSION 4
+
+
+//!TEXTURE
+Texture2D INPUT;
+
+//!SAMPLER
+//!FILTER POINT
+SamplerState sam_INPUT;
+
+
+//!TEXTURE
+//!WIDTH  INPUT_WIDTH * 2
+//!HEIGHT INPUT_HEIGHT * 2
+Texture2D OUTPUT;
+
+//!SAMPLER
+//!FILTER POINT
+SamplerState sam_OUTPUT;
+
+
+//!SAMPLER
+//!FILTER LINEAR
+SamplerState sam_INPUT_LINEAR;
+
+//!TEXTURE
+//!SOURCE ravu_lut2_f16.dds
+//!FORMAT R16G16B16A16_FLOAT
+Texture2D ravu_lut2;
+
+//!SAMPLER
+//!FILTER LINEAR
+SamplerState sam_ravu_lut2;
+
+
+
+//!TEXTURE
+//!FORMAT R16_FLOAT
+//!WIDTH  INPUT_WIDTH
+//!HEIGHT INPUT_HEIGHT
+Texture2D ravu_int11;
+
+//!SAMPLER
+//!FILTER POINT
+SamplerState sam_ravu_int11;
+
+
+
+
+//!COMMON
+// Conversion from GLSL to HLSL is done through defines as much as possible to ease synchronization and comparison with upstream
+#define ivec2 int2
+
+#define vec2 float2
+#define vec3 float3
+#define vec4 float4
+
+#define shared groupshared
+
+// TODO: check
+// some sources suggest that atan2 has reverse order of arguments compared to atan
+#define atan atan2
+#define fract frac
+#define intBitsToFloat asfloat
+#define inversesqrt rsqrt
+#define mix lerp
+
+// mod deals only with positive numbers here and it could be substituted by fmod
+#define mod fmod
+
+#define barrier GroupMemoryBarrierWithGroupSync
+#define texture(tex, pos) tex.SampleLevel(sam_##tex, pos, 0.0)
+
+// TODO: check
+// HLSL uses row-major matrixes, while GLSL uses column-major matrixes
+// Is this the proper way to deal with this difference?
+#define mat4x3 float4x3
+#define matrixCompMult(mtx1, mtx2) (mtx1 * mtx2)
+
+#define OUTPUT_pt float2(GetOutputPt())
+#define frag_pos(id) (vec2(id) + vec2(0.5, 0.5))
+#define frag_map(id) (OUTPUT_pt * frag_pos(id))
+
+#define gl_LocalInvocationIndex (threadId.y*MP_NUM_THREADS_X + threadId.x)
+#define gl_LocalInvocationID threadId
+#define gl_WorkGroupSize (uint2(MP_NUM_THREADS_X, MP_NUM_THREADS_Y))
+#define gl_WorkGroupID (blockStart / uint2(MP_BLOCK_WIDTH, MP_BLOCK_HEIGHT))
+#define gl_GlobalInvocationID (gl_WorkGroupID*gl_WorkGroupSize + threadId.xy)
```

Comment:
Please consider moving these codes to a common header file (.hlsli).

Reply:
Done.

Reply:
```hlsl
#define GET_SAMPLE(x) dot(x.rgb, rgb2y)
#define imageStore(out_image, pos, val) imageStoreOverride(pos, val.x)
void imageStoreOverride(uint2 pos, float value) {
	float2 UV = mul(rgb2uv, INPUT.SampleLevel(sam_INPUT_LINEAR, HOOKED_map(pos), 0).rgb);
	OUTPUT[pos] = float4(mul(yuv2rgb, float3(value.x, UV)), 1.0);
}

#define INPUT_tex(pos) GET_SAMPLE(vec4(texture(INPUT, pos)))
static const float2 INPUT_size = float2(GetInputSize());
static const float2 INPUT_pt = float2(GetInputPt());

#define ravu_lite_lut3_tex(pos) (vec4(texture(ravu_lite_lut3, pos)))

#define HOOKED_tex(pos) INPUT_tex(pos)
#define HOOKED_size INPUT_size
#define HOOKED_pt INPUT_pt
```
These codes can also be moved to the header file.

Reply:
These parts are generated based on different conditions inside generation scripts: [magpie.py#L243](https://github.com/hauuau/magpie-prescalers/blob/source-magpie-v4/magpie.py#L243)

I don't think it will be feasible to move them to a common header file.

Reply:
Okay, I get it. This is necessary to keeps things as close as possible to the upstream.